streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1543,26 +1544,28 @@ private boolean close(final Optional<Long> timeout, final boolean leaveGroup) {
             timeoutMs = Long.MAX_VALUE;
         }
 
-        if (state.hasCompletedShutdown()) {
-            log.info("Streams client is already in the terminal {} state, all resources are closed and the client has stopped.", state);
-            return true;
-        }
-        if (state.isShuttingDown()) {
-            log.info("Streams client is in {}, all resources are being closed and the client will be stopped.", state);
-            if (state == State.PENDING_ERROR && waitOnState(State.ERROR, timeoutMs)) {
-                log.info("Streams client stopped to ERROR completely");
-                return true;
-            } else if (state == State.PENDING_SHUTDOWN && waitOnState(State.NOT_RUNNING, timeoutMs)) {
-                log.info("Streams client stopped to NOT_RUNNING completely");
+        if (!setState(State.PENDING_SHUTDOWN)) {
+
+            if (state.isShuttingDown()) {
+                log.info("Streams client is in {}, all resources are being closed and the client will be stopped.", state);
+                if (state == State.PENDING_ERROR && waitOnState(State.ERROR, timeoutMs)) {
```

Comment:
Not sure logic is right? (I know you just move it, but it was not right to begin with)

`setState(PENDING_SHUTDOWN)` above is an atomic compare-and-set as we get a lock. So here, we only know we are in one of four states,` PENDING_SHUTDOWN`, `NOT_RUNNING`, `PENDING_ERROR` or `ERROR`.

After `isShuttingDown()` returns, state could have transit to `ERROR` before be get here (similar to `PENDING_SHUTDOWN` below; we could be in `NOT_RUNNING` already), so we would end up in the final `else` and log a WARN and return `false` incorrectly.

Should we combine both conditions into a single `waitOnState(ERROR || NOT_RUNNING)` (need a new `waitOnState` overload that take two "target states"), and change the log to `log.info("Streams client stopped to {} completely", state);`?